rhino/src/main/java/org/mozilla/javascript/Context.java

```diff
@@ -528,28 +535,32 @@ static final Context enter(Context cx, ContextFactory factory) {
      * @see ContextFactory#enterContext()
      */
     public static void exit() {
-        Object helper = VMBridge.instance.getThreadContextHelper();
-        Context cx = VMBridge.instance.getContext(helper);
+        Context cx = currentContext.get();
         if (cx == null) {
             throw new IllegalStateException("Calling Context.exit without previous Context.enter");
         }
         if (cx.enterCount < 1) Kit.codeBug();
         if (--cx.enterCount == 0) {
-            VMBridge.instance.setContext(helper, null);
-            cx.factory.onContextReleased(cx);
+            releaseContext(cx);
         }
     }
 
     @Override
     public void close() {
         if (enterCount < 1) Kit.codeBug();
         if (--enterCount == 0) {
-            Object helper = VMBridge.instance.getThreadContextHelper();
-            VMBridge.instance.setContext(helper, null);
-            factory.onContextReleased(this);
+            if (currentContext.get() != this) Kit.codeBug();
```

Comment:
please check, if this would make sense

Reply:
I don't think it's bad to check but it has the potential to break existing code. How about instead:

`assert currentContext.get() == null`